### PR TITLE
update electron to version 11

### LINKED ISSUE #9095: Update to Electron 11
### Feature Description:
In order to build Theia for Apple M1 chip electron need to be updated to version 11.
Currently, Theia can run on Apple M1 chip thanks to Rosetta.
But with native build for Apple M1 chip the performance will increase.

### PR DESCRIPTION
**Pull request by linhanyu**
update electron-store to version 7. Its required by electron 11.
update node to version 12.18.3. Its required by electron 11.
update docs with minimum version of node.
update minimum version of node in project.
update node-pty to 0.10.0. It's required by electron-rebuild
fix type error

Closes #9095.

Signed-off-by: 翰宇 林 <305991071@qq.com>

#### What it does

update electron to version 11
update electron-store to version 7. Its required by electron 11.
update node to version 12.18.3. Its required by electron 11.
update docs with minimum version of node.
update minimum version of node in project.
update node-pty to 0.10.0. It's required by electron-rebuild
fix compile error

Closes #9095.

#### How to test

  Pull changes from repository branch.
    Install node 12.18.3
    Install dependencies by running command yarn in project root directory.
    Run theia electron example by running command yarn start:electron
    In case of errors caused by native modules run command yarn
    rebuild:clean && yarn rebuild:electron and repeat previous step.
    Open chrome dev tools execute process.versions and verify that electron
    version is 11.
    Manually test functionalities.

    Use electron-builder, electron-rebuild to rebuild all native dependency and checkout those native module functionalities
#### Review checklist

- [ ] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/core/src/electron-browser/menu/electron-main-menu-factory.ts [hunk 1]
```diff
@@ -74,7 +74,11 @@ export class ElectronMainMenuFactory {
         preferencesService.onPreferenceChanged(debounce(() => {
             if (this._menu) {
                 for (const item of this._toggledCommands) {
-                    this._menu.getMenuItemById(item).checked = this.commandRegistry.isToggled(item);
+                    const menu = this._menu.getMenuItemById(item);
+                    if (!menu) {
+                        continue;
+                    }
+                    menu.checked = this.commandRegistry.isToggled(item);
```
Comment (msujew) [inline packages/core/src/electron-browser/menu/electron-main-menu-factory.ts:81]: Nitpick: You could invert the `!menu` check and run `menu.checked = this.commandRegistry.isToggled(item);` inside the branch, instead of introducing mental overhead with `continue`. It looks like you already did that in line 270.

File: CHANGELOG.md [hunk 2]
```diff
@@ -651,12 +651,12 @@
 <a name="1_5_0_drop_node_10_support"></a>
 - [[repo]](#1_5_0_drop_node_10_support) support for `Node 10` is dropped. [#8290](https://github.com/eclipse-theia/theia/pull/8290)
   - From now on, Node.js `12.x` is required when building.\
-   The recommended minimum version is aligned with `electron` (Node.js `12.14.1`).
+   The recommended minimum version is aligned with `electron` (Node.js `12.18.3`).
```
Comment (msujew) [inline CHANGELOG.md:654]: You shouldn't update old changelogs. Instead, add a breaking change section to version `1.17.0`

File: CHANGELOG.md [hunk 3]
```diff
@@ -651,12 +651,12 @@
...
 
 ## v1.4.0 - 30/07/2020
 
 - [core] added support for Node.js `12.x` [#7968](https://github.com/eclipse-theia/theia/pull/7968)
-  - From now on, you can use Node.js `12.x` to build Theia from sources. The recommended minimum version is aligned with `electron` (Node.js `12.14.1`).
+  - From now on, you can use Node.js `12.x` to build Theia from sources. The recommended minimum version is aligned with `electron` (Node.js `12.18.3`).
```
Comment (msujew) [inline CHANGELOG.md:659]: You shouldn't update old changelogs. Please revert these changes.

File: .gitpod.dockerfile [hunk 4]
```diff
@@ -5,22 +5,22 @@ ENV TRIGGER_REBUILD 3
...
-    && sudo rm -rf /var/lib/apt/lists/*
+  # window manager
+  && sudo apt-get install -y jwm \
+  # electron
+  && sudo apt-get install -y libgtk-3-0 libnss3 libasound2 libgbm1 \
+  # native-keymap
+  && sudo apt-get install -y libx11-dev libxkbfile-dev \
+  # keytar
+  && sudo apt-get install -y libsecret-1-dev \
+  && sudo rm -rf /var/lib/apt/lists/*
```
Comment (msujew) [inline .gitpod.dockerfile:16]: It looks like your IDE autoformatted this code. Would you mind reverting the formatting changes? We usually use a tab size of 4 spaces everywhere.

File: packages/core/src/electron-browser/menu/electron-main-menu-factory.ts [hunk 5]
```diff
@@ -265,7 +269,11 @@ export class ElectronMainMenuFactory {
             if (this.commandRegistry.isEnabled(command, ...args)) {
                 await this.commandRegistry.executeCommand(command, ...args);
                 if (this._menu && this.commandRegistry.isVisible(command, ...args)) {
-                    this._menu.getMenuItemById(command).checked = this.commandRegistry.isToggled(command, ...args);
+                    const menu = this._menu.getMenuItemById(command);
+                    if (menu) {
+
```
Comment (msujew) [inline packages/core/src/electron-browser/menu/electron-main-menu-factory.ts:274]: Nitpick: Would you mind removing the empty line here?

**Review by msujew — COMMENTED** [review]
Thank you for your contribution, that already looks really well. I was able to build the application on Windows and ran some manual smoke tests, especially regarding the terminal. I couldn't notice any regressions.

I have a few nitpicks though. Also hadn't had the time to actually build an electron deployment yet, which I will also try to do.

**Comment by linhanyu** [comment]
thanks，I had fixed those problem
What's more, today I have build this branch in my m1 laptop. Electron build failed for the ffmpeg subtitute script( fail in dlopen ),because the ffmpeg.node library were build in x86_64 but not arm64.

**Comment by westbury** [comment]
@linhanyu thank you for this PR.  I assume you are aware of https://github.com/eclipse-theia/theia/pull/9096 and that this is intended to replace that PR.  I hope to get some time to try this out on my M1 soon.

**Comment by linhanyu** [comment]
Yes，the difference was I had&nbsp; replaced @theia/node-pty with node-pty and fix some compile error.Then test functionality on my windows, and macos laptop.





---Original---
From: "Nigel ***@***.***&gt;
Date: Mon, Aug 2, 2021 20:42 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [eclipse-theia/theia] update electron to version 11 (#9813)




 
@linhanyu thank you for this PR.  I assume you are aware of #9096 and that this is intended to replace that PR.  I hope to get some time to try this out on my M1 soon.

**Comment by paul-marechal** [comment]
Closing in favor of https://github.com/eclipse-theia/theia/pull/9936

Please share your feedback there if any.
